asset/asset.go

```diff
@@ -520,23 +523,43 @@ func NewRawKeyGenesisSigner(priv *btcec.PrivateKey) *RawKeyGenesisSigner {
 	}
 }
 
-// SignGenesis signs the passed Genesis description using the public key
-// identified by the passed key descriptor. The final tweaked public key and
-// the signature are returned.
+// SignGenesis tweaks the public key identified by the passed key
+// descriptor with the the first passed Genesis description, and signs
+// the second passed Genesis description with the tweaked public key.
+// For minting the first asset in a group, only one Genesis object is
+// needed, since we tweak with and sign over the same Genesis object.
+// The final tweaked public key and the signature are returned.
 func (r *RawKeyGenesisSigner) SignGenesis(keyDesc keychain.KeyDescriptor,
-	gen Genesis) (*btcec.PublicKey, *schnorr.Signature, error) {
+	initialGen Genesis, currentGen *Genesis) (*btcec.PublicKey,
+	*schnorr.Signature, error) {
 
 	if !keyDesc.PubKey.IsEqual(r.privKey.PubKey()) {
 		return nil, nil, fmt.Errorf("cannot sign with key")
 	}
 
 	tweakedPrivKey := txscript.TweakTaprootPrivKey(
-		*r.privKey, gen.GroupKeyTweak(),
+		*r.privKey, initialGen.GroupKeyTweak(),
 	)
 
+	// If the current genesis is not set, we are minting the first asset in
+	// the group. This means that we use the same Genesis object for both
+	// the key tweak and to create the asset ID we sign. If the current
+	// genesis is set, the asset type of the new asset must match the type
+	// of the first asset in the group.
+	id := initialGen.ID()
+	if currentGen != nil {
+		if initialGen.Type != currentGen.Type {
+			return nil, nil, fmt.Errorf(
```

Comment:
nit (non-blocking, feel free to merge without fixing): formatting of `fmt.Errorf()` can be compact form, here and below.